### PR TITLE
Add generic attributes

### LINKED ISSUE #25681: Generic attributes
[Feature specification](https://github.com/dotnet/csharplang/blob/main/proposals/csharp-10.0/generic-attributes.md)

Generic classes can now specify `System.Attribute` as a base class.

### PR DESCRIPTION
**Pull request by BillWagner**
Fixes dotnet/roslyn#25681

Add a description of the generic attributes preview feature.

### CODE REVIEW

File: docs/csharp/whats-new/csharp-10.md [hunk 1]
```diff
@@ -168,3 +169,58 @@ You can learn more about this feature in the article on [Caller information attr
...
+{
+   [GenericAttribute<T>()] // Not allowed! generic attributes must be fully closed types.
+   public string Method() => default;
+}
+```
+
+The type arguments must satisfy the same restrictions as the [`typeof`](../language-reference/operators/type-testing-and-cast.md#typeof-operator) operator. Types that require metadata annotations aren't allowed. Examples include the following:
+
+- `dynanmic`
+- `nint`, `nuint`
```
Comment (BillWagner) [inline docs/csharp/whats-new/csharp-10.md:222]: @RikkiGibson These two work with `typeof`. Should they work here?

File: docs/csharp/whats-new/csharp-10.md [hunk 2]
```diff
@@ -168,3 +169,58 @@ You can learn more about this feature in the article on [Caller information attr
...
+- `object?` (or any nullable reference type)
+- `(int X, int Y)` (or any other tuple types using C# tuple syntax). Note that you can use the underlying type of `ValueTuple<int, int>`.
```
Comment (BillWagner) [inline docs/csharp/whats-new/csharp-10.md:224]: @RikkiGibson This also works with `typeof`. Same question.

File: docs/csharp/whats-new/csharp-10.md (hunk 1 above)
Comment (RikkiGibson) [reply docs/csharp/whats-new/csharp-10.md:222]: What happened here is: we took generic attributes as an opportunity to "tighten up" the rules on using type arguments that can't actually be represented in metadata. However, for compat, it proved to be too difficult at least for .NET 6 to bring the same rules back over to typeof, even though typeof has similar limitations as generic attribute type arguments.

Related discussion: dotnet/roslyn#54804

Comment (BillWagner) [reply docs/csharp/whats-new/csharp-10.md:222]: Thanks @RikkiGibson That discussion helped. I added some clarifications and what types to use instead.

**Comment by BillWagner** [comment]
> This is ready, pending the questions for the compiler team.

This is ready now.

File: docs/csharp/language-reference/operators/type-testing-and-cast.md [hunk 3]
```diff
@@ -121,23 +121,30 @@ The `typeof` operator obtains the <xref:System.Type?displayProperty=nameWithType
 
 [!code-csharp-interactive[typeof operator](snippets/shared/TypeTestingAndConversionOperators.cs#TypeOf)]
 
+The argument mustn't be a type that requires metadata annotations. Examples include the following types:
+
+- `dynanmic`
```
Comment (gewarren) [inline docs/csharp/language-reference/operators/type-testing-and-cast.md:126]:
```suggestion
- `dynamic`
```

File: docs/csharp/whats-new/csharp-10.md [hunk 4]
```diff
@@ -22,7 +22,11 @@ C# 10 adds the following features and enhancements to the C# language:
 - [CallerArgumentExpression attribute](#callerargumentexpression-attribute-diagnostics)
 - [Enhanced `#line` pragma](#enhanced-line-pragma)
 
-Some of the features you can try are available only when you set your language version to "preview". These features may have more refinements in future previews before .NET 6 is released.
+In addition, there are two features that are available in *preview* mode. You're encouraged to try these features and provide feedback on them. They may change before their final release. In order to use these features, you must [set the `<LangVersion>` to `Preview`](../language-reference/compiler-options/language.md#langversion) in your project.
```
Comment (gewarren) [inline docs/csharp/whats-new/csharp-10.md:25]:
```suggestion
In addition, there are two features that are available in *preview* mode. You're encouraged to try these features and provide feedback on them. They may change before their final release. In order to use these features, you must [set `<LangVersion>` to `Preview`](../language-reference/compiler-options/language.md#langversion) in your project.
```

File: docs/csharp/whats-new/csharp-10.md [hunk 5]
```diff
@@ -168,3 +169,63 @@ You can learn more about this feature in the article on [Caller information attr
 ## Enhanced #line pragma
 
 C# 10 supports a new format for the `#line` pragma. You likely won't use the new format, but you'll see its effects. The enhancements enable more fine-grained output in domain-specific languages (DSLs) like Razor. The Razor engine uses these enhancements to improve the debugging experience. You'll find debuggers can highlight your Razor source more accurately. To learn more about the new syntax, see the article on [Preprocessor directives](../language-reference/preprocessor-directives.md#error-and-warning-information) in the language reference. You can also read the [feature specification](~/_csharplang/proposals/csharp-10.0/enhanced-line-directives.md#examples) for Razor based examples.
+
+## Generic attributes
+
+> [!IMPORTANT]
+> *Generic attributes* is a preview feature. You must [set the `<LangVersion>` to `Preview`](../language-reference/compiler-options/language.md#langversion) to enable this feature. This feature may change before its final release.
```
Comment (gewarren) [inline docs/csharp/whats-new/csharp-10.md:176]:
```suggestion
> *Generic attributes* is a preview feature. You must [set `<LangVersion>` to `Preview`](../language-reference/compiler-options/language.md#langversion) to enable this feature. This feature may change before its final release.
```

File: docs/csharp/whats-new/csharp-10.md [hunk 6]
```diff
@@ -168,3 +169,63 @@ You can learn more about this feature in the article on [Caller information attr
...
+
+You can declare a [generic class](../programming-guide/generics/generic-classes.md) whose base class is <xref:System.Attribute?displayProperty=fullName>. This provides a more convenient syntax for atributes that require a <xref:System.Type?displayProperty=nameWithType> parameter. Previously, you'd need to write an attribute that takes a `Type` as its parameter:
```
Comment (gewarren) [inline docs/csharp/whats-new/csharp-10.md:178]:
```suggestion
You can declare a [generic class](../programming-guide/generics/generic-classes.md) whose base class is <xref:System.Attribute?displayProperty=fullName>. This provides a more convenient syntax for attributes that require a <xref:System.Type?displayProperty=nameWithType> parameter. Previously, you'd need to create an attribute that takes a `Type` as its constructor parameter:
```

File: docs/csharp/whats-new/csharp-10.md [hunk 7]
```diff
@@ -168,3 +169,63 @@ You can learn more about this feature in the article on [Caller information attr
...
+```csharp
+public class TypeAttribute : Attribute
+{
+   public TypeAttribute(Type t) => ParamType = t;
+
+   public Type ParamType { get; }
+}
+```
+
+To apply the attribute, you use the [`typeof`](../language-reference/operators/type-testing-and-cast.md#typeof-operator) operator:
```
Comment (gewarren) [inline docs/csharp/whats-new/csharp-10.md:189]:
```suggestion
And to apply the attribute, you use the [`typeof`](../language-reference/operators/type-testing-and-cast.md#typeof-operator) operator:
```

File: docs/csharp/whats-new/csharp-10.md [hunk 8]
```diff
@@ -168,3 +169,63 @@ You can learn more about this feature in the article on [Caller information attr
...
+
+```csharp
+[TypeAttribute(typeof(string))] 
+public string Method() => default;
+```
+
+Using this feature, you can use a generic attribute instead:
```
Comment (gewarren) [inline docs/csharp/whats-new/csharp-10.md:196]:
```suggestion
Using this new feature, you can create a generic attribute instead:
```

File: docs/csharp/whats-new/csharp-10.md [hunk 9]
```diff
@@ -168,3 +169,63 @@ You can learn more about this feature in the article on [Caller information attr
...
+
+- `dynamic`
+- `nint`, `nuint`
+- `string?` (or any nullable reference type)
+- `(int X, int Y)` (or any other tuple types using C# tuple syntax).
+
+These types aren't directly represented in metadata. They types include annotations that describe the type. In all cases, you can use the underlying type instead:
+
+- `object` for `dynamic`.
+- <xref:System.IntPtr> instead of `nint`, or `unint`.
```
Comment (gewarren) [inline docs/csharp/whats-new/csharp-10.md:229]:
```suggestion
- <xref:System.IntPtr> instead of `nint` or `unint`.
```

**Review by gewarren — APPROVED** [review]
Some minor suggestions to consider.

File: docs/csharp/whats-new/csharp-10.md [hunk 10]
```diff
@@ -22,7 +22,11 @@ C# 10 adds the following features and enhancements to the C# language:
 - [CallerArgumentExpression attribute](#callerargumentexpression-attribute-diagnostics)
 - [Enhanced `#line` pragma](#enhanced-line-pragma)
 
-Some of the features you can try are available only when you set your language version to "preview". These features may have more refinements in future previews before .NET 6 is released.
+In addition, there are two features that are available in *preview* mode. You're encouraged to try these features and provide feedback on them. They may change before their final release. In order to use these features, you must [set the `<LangVersion>` to `Preview`](../language-reference/compiler-options/language.md#langversion) in your project.
+
+The features available only in preview mode are:
+
+- [Generic attributes](#generic-attributes)
```
Comment (gewarren) [inline docs/csharp/whats-new/csharp-10.md:29]: Is something missing here? Perhaps CallerArgumentExpression?

Comment (BillWagner) [reply docs/csharp/whats-new/csharp-10.md:29]: static abstract interfaces are coming in my next PR. I'll reword this for now.

**Pull request merged**